File: game-server/src/main/content/global/skill/summoning/familiar/FamiliarItemOptionPlugin.java
package content.global.skill.summoning.familiar;

import content.global.skill.summoning.pet.Pets;
import core.cache.def.impl.ItemDefinition;
import core.game.interaction.OptionHandler;
import core.game.node.Node;
import core.game.node.entity.player.Player;
import core.game.node.item.Item;
import core.plugin.Initializable;
import core.plugin.Plugin;

/**
 * The type Familiar item option plugin.
 */
@Initializable
public final class FamiliarItemOptionPlugin extends OptionHandler {

    @Override
    public Plugin<Object> newInstance(Object arg) throws Throwable {
        for (Pets p : Pets.values()) {
            ItemDefinition def = ItemDefinition.forId(p.babyItemId);
            if (def == null) {
                continue;
            }
            def.getHandlers().put("option:drop", this);
            def.getHandlers().put("option:release", this);
            if (p.grownItemId > -1) {
                ItemDefinition.forId(p.grownItemId).getHandlers().put("option:drop", this);
                ItemDefinition.forId(p.grownItemId).getHandlers().put("option:release", this);
            }
            if (p.overgrownItemId > -1) {
                ItemDefinition.forId(p.overgrownItemId).getHandlers().put("option:drop", this);
                ItemDefinition.forId(p.overgrownItemId).getHandlers().put("option:release", this);
            }
        }
        return this;
    }

    @Override
    public boolean handle(Player player, Node node, String option) {
        switch (option) {
            case "drop":
                player.getFamiliarManager().summon(((Item) node), true);
                return true;
            case "release":
                if (node.getId() == 7771) {
                    player.getFamiliarManager().summon(((Item) node), true);
                    return true;
                }
                if (player.getInventory().remove(((Item) node))) {
                    player.getDialogueInterpreter().sendDialogues(player, null, "Run along; I'm setting you free.");
                }
                return true;
        }
        return true;
    }

    @Override
    public boolean isWalk() {
        return false;
    }
}

File: Server/src/main/content/global/skill/summoning/familiar/FamiliarItemOptionPlugin.kt
package content.global.skill.summoning.familiar

import content.global.skill.summoning.pet.Pets
import core.cache.def.impl.ItemDefinition.Companion.forId
import core.game.interaction.OptionHandler
import core.game.node.Node
import core.game.node.entity.player.Player
import core.game.node.item.Item
import core.plugin.Initializable
import core.plugin.Plugin
import shared.consts.Items

/**
 * The familiar item option plugin.
 */
@Initializable
class FamiliarItemOptionPlugin : OptionHandler() {
    @Throws(Throwable::class)
    override fun newInstance(arg: Any?): Plugin<Any> {
        for (p in Pets.values()) {
            val def = forId(p.babyItemId) ?: continue
            def.handlers["option:drop"] = this
            def.handlers["option:release"] = this
            if (p.grownItemId > -1) {
                forId(p.grownItemId).handlers["option:drop"] = this
                forId(p.grownItemId).handlers["option:release"] = this
            }
            if (p.overgrownItemId > -1) {
                forId(p.overgrownItemId).handlers["option:drop"] = this
                forId(p.overgrownItemId).handlers["option:release"] = this
            }
        }
        return this
    }

    override fun handle(player: Player, node: Node, option: String): Boolean {
        when (option) {
            "drop" -> {
                player.familiarManager.summon((node as Item), true)
                return true
            }

            "release" -> {
                if (node.id == Items.CLOCKWORK_CAT_7771) {
                    player.familiarManager.summon((node as Item), true)
                    return true
                }
                if (player.inventory.remove((node as Item))) {
                    player.dialogueInterpreter.sendDialogues(player, null, "Run along; I'm setting you free.")
                }
                return true
            }
        }
        return true
    }

    override fun isWalk(): Boolean {
        return false
    }
}
